### PR TITLE
Correct refreshing logic of pre-signed urls for delta sharing streaming

### PR DESCRIPTION
**Pull request by linzhou-db**
Correct refreshing logic of pre-signed urls for delta sharing streaming:
1. refresh immediately if it already expired.
2. set the correct expiration time for a CachedTable 

Tested with a integration test from a notebook.

### CODE REVIEW

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala:246]: As discussed, we should update the sortedFetchedFiles as well in order to avoid redundant refreshes.

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala:152]: added.

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala]: updated to outside.

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala]: updated.

**Review by zsxwing — APPROVED** [review]
LGTM

File: spark/src/main/scala/org/apache/spark/delta/sharing/PreSignedUrlCache.scala
Comment (linzhou-db) [reply spark/src/main/scala/org/apache/spark/delta/sharing/PreSignedUrlCache.scala:146]: done.

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala:511]: Question: This seems like an on-demand refresh? Will we do double refresh... in CacheTable, and here?

Comment (chakankardb) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala:511]: As discussed, let's optimize this in a follow up PR.

Also, add a flag to protect this.

Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala:515]: added a flag, please take a look. Thanks!

**Review by chakankardb — APPROVED** [review]

**Comment by linzhou-db** [comment]
Tested again with the flag, will merge once all pre-merge tests are done.

**Pull request merged**
